### PR TITLE
perf: avoid reproposed uncle proposals

### PR DESCRIPTION
**Pull request by keroro520**
* test: Add avoid_duplicated_proposals_with_uncles
* perf: Avoid re-proposed uncle transactions

### CODE REVIEW

File: tx-pool/src/pool.rs [hunk 1]
```diff
@@ -664,10 +664,14 @@ impl TxPool {
         ret
     }
 
-    pub fn get_proposals(&self, limit: usize) -> HashSet<ProposalShortId> {
+    pub fn get_proposals(
+        &self,
+        limit: usize,
+        filter: &HashSet<ProposalShortId>,
```
Comment (doitian) [inline tx-pool/src/pool.rs]: Use `exclusion`, filter is ambiguous.

File: tx-pool/src/component/pending.rs [hunk 2]
```diff
@@ -56,11 +56,16 @@ impl PendingQueue {
     }
 
     // fill proposal txs
-    pub fn fill_proposals(&self, limit: usize, proposals: &mut HashSet<ProposalShortId>) {
+    pub fn fill_proposals(
+        &self,
+        limit: usize,
+        filter: &HashSet<ProposalShortId>,
```
Comment (doitian) [inline tx-pool/src/component/pending.rs]: Use `exclusion`, filter is ambiguous.

File: tx-pool/src/component/pending.rs [hunk 3]
```diff
@@ -56,11 +56,16 @@ impl PendingQueue {
...
+        proposals: &mut HashSet<ProposalShortId>,
+    ) {
         for id in self.sorted_keys() {
             if proposals.len() == limit {
                 break;
-            } else if proposals.contains(&id) {
+            } else if filter.contains(&id) || proposals.contains(&id) {
                 // implies that ancestors are already in proposals
                 continue;
             }
```
Comment (u2) [inline tx-pool/src/component/pending.rs]: If the proposal is in uncle already, we should continue to handle the ancestor of it, not just skip simply?

Comment (keroro520) [reply tx-pool/src/component/pending.rs]: You are right. We should continue to handle its ancestors.
I will fix it and try to add a unit test soon.

Comment (keroro520) [reply tx-pool/src/component/pending.rs]: Adding a unit test to cover `handle its ancestors` is out of this PR.

File: tx-pool/src/component/pending.rs [hunk 4]
```diff
@@ -71,7 +76,15 @@ impl PendingQueue {
                     .expect("exists")
             });
             ancestors.push(id.clone());
-            proposals.extend(ancestors.into_iter().take(limit - proposals.len()));
```
Comment (keroro520) [inline tx-pool/src/component/pending.rs]: In fact, this previous line has a small bug: Consider the below case,

```
limit = 2
proposals = [a]
ancestors = [a, b, c, d]
```

The final `proposals` by previous code is `[a]`, but what we expect is `[a, b]`

**Comment by u2** [comment]
@doitian

**Comment by keroro520** [comment]
@doitian Please review again.

**Comment by quake** [comment]
@keroro520 please rebase

**Comment by doitian** [comment]
Please resolve conflicts

**Comment by zhangsoledad** [comment]
bors r=quake,zhangsoledad

**Pull request merged**